core/src/main/java/io/wttech/stubway/StubFinderService.java

```diff
@@ -37,38 +37,34 @@ public StubResponse getStubResponse(SlingHttpServletRequest request) {
 	}
 
 	private StubResponse createStubResponse(Resource stubsResource, RequestParameters params) {
-		for (Stub stub : getAllStubs(stubsResource)) {
-			try {
-				if (isStubMatching(stub, params) && isMethodMatching(stub, params)) {
-					return StubResponse.foundStub(stub);
-				}
-			} catch (MissingSupportedMethodException exception) {
-				return StubResponse.error(exception);
-			}
-		}
-		return StubResponse.empty();
+		return getStub(stubsResource, params).map(StubResponse::foundStub).orElse(StubResponse.empty());
 	}
 
-	private List<Stub> getAllStubs(Resource stubsResource) {
+	private Optional<Stub> getStub(Resource stubsResource, RequestParameters params) {
 		return Optional.ofNullable(stubsResource)
 				.map(Resource::getChildren)
-				.map(this::toStubs)
-				.orElse(Collections.emptyList());
+				.map(resources -> toStub(resources, params));
 	}
 
-	private List<Stub> toStubs(Iterable<Resource> resources) {
+	private Stub toStub(Iterable<Resource> resources, RequestParameters params) {
 		return StreamSupport.stream(resources.spliterator(), false)
 				.map(resource -> resource.adaptTo(Stub.class))
 				.filter(Objects::nonNull)
-				.collect(Collectors.toList());
+				.filter(stub -> isMethodMatching(stub, params))
+				.filter(stub -> isStubMatching(stub, params))
+				.findFirst()
+				.orElse(null);
 	}
 
-	private boolean isMethodMatching(Stub stub, RequestParameters params) throws MissingSupportedMethodException {
-		return ObjectUtils.equals(params.getMethod(), stub.getMethod());
+	private boolean isMethodMatching(Stub stub, RequestParameters params) {
+		try {
+			return ObjectUtils.equals(params.getMethod(), stub.getMethod());
+		} catch (MissingSupportedMethodException ex) {
```

Comment:
I moved the exception handling to this method for 2 reasons:
- In practice, requests with unsupported HTTP methods were not really reaching this code. It seems they are already treated elsewhere and I've verified that the specific error message for unsupported method is still returned whenever any method other than "GET" or "SET" is used after my changes.
- I think it actually makes more sense to treat this case as a false match at this point, due to the nature of the comparison we're doing here.

As a bonus, propagating the exception to the other methods would add a lot of complication to the code with no actual benefit (given the point in the first bullet)